### PR TITLE
feat(query): hash join adaptive early filtering

### PR DESCRIPTION
**Pull request by Dousir9**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

### 1. Adaptive early filtering

截屏2023-11-06 19 18 50

For each bucket in the hash table, its type is u64. We use the upper 16 bits to store a tag for early filtering and the lower 48 bits to store a pointer to the `RowEntry`. We construct the tag during the finalize phase of the hash join, encoding it into the upper 16 bits. During the probing phase, we can check the tag first. If the key is not found in the tag, we can avoid reading the `RowEntry`, reducing our random memory accesses. Through testing, we have observed an average **25%** reduction in memory accesses in TPC-H.
Thanks to the adaptive execution strategy of early filtering, we also don't experience a performance decrease when all keys have matches. This allows us to achieve the same performance as before.

### 2. Improve the readability and maintainability of the hash join code
Early filtering introduces selection to iterate keys. For each type of join, by introducing `BuildState` and `ProbeState`, we extract common code into functions like `process_xxx_block`. For example, for inner join, we extract common code into `process_inner_block`, and for right join, we extract it into `process_right_block`. This improvement enhances code readability and maintainability.

### 3. Benchmark TPC-H SF100
Take the minimum execution time out of 5 runs for each query: 
| query | main      | pr        | improvement |
| ----- | --------- | --------- | ----------- |
| Q9    | 9.525181  | 8.473911  | 12%         |
| Q22   | 1.130324  | 1.015136  | 11%         |
| Q3    | 4.353114  | 3.959412  | 10%         |
| Q4    | 2.973388  | 2.713897  | 10%         |
| Q8    | 4.61047   | 4.197131  | 10%         |
| Q21   | 8.546927  | 7.758365  | 10%         |
| Q5    | 5.577217  | 5.128655  | 9%          |
| Q7    | 4.43625   | 4.232459  | 5%          |
| Q16   | 1.140485  | 1.087739  | 5%          |
| Q20   | 14.092364 | 13.597196 | 4%          |
| Q10   | 5.429098  | 5.277602  | 3%          |
| Q12   | 5.017574  | 4.863078  | 3%          |
| Q13   | 5.292683  | 5.12285   | 3%          |
| Q18   | 17.038875 | 16.619233 | 3%          |
| Q14   | 2.768227  | 2.721796  | 2%          |
| Q19   | 7.21885   | 7.04369   | 2%          |
| Q2    | 1.461207  | 1.442259  | 1%          |
| Q17   | 16.945474 | 16.708737 | 1%          |
| Q1    | 9.318129  | 9.338366  | 0%          |
| Q6    | 2.643723  | 2.651717  | 0%          |
| Q11   | 0.900256  | 0.897974  | 0%          |
| Q15   | 4.849567  | 4.83456   | 0%          |

#### Q3 before:
截屏2023-11-17 19 07 50

#### Q3 after
截屏2023-11-17 19 08 30

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13393)

### CODE REVIEW

**Comment by BohuTANG** [comment]
Great to hear the perfmance improved.
But this is a monster PR, suggest spilt into many smalls and each PR we can test and measure.
It's hard to maintain the RISK.

**Comment by Dousir9** [comment]
> Great to hear the perfmance improved. But this is a monster PR, suggest spilt into many smalls and each PR we can test and measure, or we add enough unit tests for the huge PR, I think this is a hard work. It's hard to maintain the RISK, this PR is on the core path of Databend.

In fact, most of the code changes are aimed at improving readability and maintainability after adding `selections`, the execution logic of the hash join remains the same as before. 😊

**Comment by BohuTANG** [comment]
> > Great to hear the perfmance improved. But this is a monster PR, suggest spilt into many smalls and each PR we can test and measure, or we add enough unit tests for the huge PR, I think this is a hard work. It's hard to maintain the RISK, this PR is on the core path of Databend.
> 
> In fact, most of the code changes are aimed at improving readability and maintainability after adding `selections`, the execution logic of the hash join remains the same as before. 😊

We can put the improve to a single PR and review, also we can add more testing for it if it not have.

File: src/common/hashtable/src/dictionary_string_hashtable.rs [hunk 1]
```diff
@@ -46,20 +46,41 @@ impl PartialEq for DictionaryKeys {
...
+                        .map(|x| x.as_ref().fast_hash())
+                        .reduce(|left, right| {
+                            let mut a = (left ^ right).wrapping_mul(0x9ddfea08eb382d69_u64);
+                            a ^= a >> 47;
+                            let mut b = (right ^ a).wrapping_mul(0x9ddfea08eb382d69_u64);
+                            b ^= b >> 47;
+                            b.wrapping_mul(0x9ddfea08eb382d69_u64)
+                        })
+                        .unwrap_or_default();
+                    hash >> 32
```
Comment (sundy-li) [inline src/common/hashtable/src/dictionary_string_hashtable.rs:66]:
```
if #[cfg(target_feature = "sse4.2")] {
    hash >> 32
} else {
   hash
}
```

**Comment by Dousir9** [comment]
> With the enhancement of join sqllogictests(#13619, #13679, #13695), we can now proceed to advance this work :) @BohuTANG @sundy-li @xudong963

File: src/common/hashtable/src/hashjoin_hashtable.rs [hunk 2]
```diff
@@ -108,10 +162,144 @@ where
...
+        if let Some(bitmap) = bitmap {
+            if bitmap.unset_bits() == bitmap.len() {
+                hashes.iter_mut().for_each(|hash| {
+                    *hash = 0;
+                });
+                return 0;
+            } else if bitmap.unset_bits() > 0 {
+                valids = Some(bitmap);
+            }
+        }
```
Comment (xudong963) [inline src/common/hashtable/src/hashjoin_hashtable.rs:176]: How about

```rust
if let Some(bitmap) = bitmap {
    if bitmap.unset_bits() == bitmap.len() {
        hashes.iter_mut().for_each(|hash| {
            *hash = 0;
        });
        return 0;
}
....
match bitmap {

}               
```

and we can extract a helper function for the code snippet
```rust
  let header = self.pointers[(*hash >> self.hash_shift) as usize];
  if header != 0 {
      *hash = remove_header_tag(header);
      count += 1;
  } else {
      *hash = 0;
  }
```

File: src/common/hashtable/src/hashjoin_hashtable.rs [hunk 3]
```diff
@@ -108,10 +162,144 @@ where
...
+                        *hash = 0;
+                    }
+                });
+            }
+        }
+        count
+    }
+
+    // Using hashes to probe hash table and converting them in-place to pointers for memory reuse.
+    fn early_filtering_probe(&self, hashes: &mut [u64], bitmap: Option<Bitmap>) -> usize {
```
Comment (xudong963) [inline src/common/hashtable/src/hashjoin_hashtable.rs:213]: ditto

File: src/common/hashtable/src/traits.rs [hunk 4]
```diff
@@ -470,7 +507,18 @@ pub trait HashJoinHashtableLike {
     type Key: ?Sized;
 
     // Using hashes to probe hash table and converting them in-place to pointers for memory reuse.
-    fn probe(&self, hashes: &mut [u64]);
+    fn probe(&self, hashes: &mut [u64], bitmap: Option<Bitmap>) -> usize;
+
+    // Using hashes to probe hash table and converting them in-place to pointers for memory reuse.
+    fn early_filtering_probe(&self, hashes: &mut [u64], bitmap: Option<Bitmap>) -> usize;
+
+    // Using hashes to probe hash table and converting them in-place to pointers for memory reuse.
```
Comment (xudong963) [inline src/common/hashtable/src/traits.rs:515]: We can make the notes more meaningful to distinguish the three methods;

File: src/query/service/src/pipelines/processors/transforms/hash_join/build_spill/build_spill_state.rs [hunk 5]
```diff
@@ -85,10 +85,12 @@ impl BuildSpillState {
...
         // Collect rows in `Chunks`
-        let chunks = unsafe { &mut *self.build_state.hash_join_state.chunks.get() };
+        let chunks = &mut unsafe { &mut *self.build_state.hash_join_state.build_state.get() }
+            .generation_state
+            .chunks;
         blocks.append(chunks);
-        let build_num_rows = unsafe { &mut *self.build_state.hash_join_state.build_num_rows.get() };
-        *build_num_rows = 0;
+        let build_state = unsafe { &mut *self.build_state.hash_join_state.build_state.get() };
+        build_state.generation_state.build_num_rows = 0;
```
Comment (xudong963) [inline src/query/service/src/pipelines/processors/transforms/hash_join/build_spill/build_spill_state.rs:93]: Seems we can encapsulate the code snippet to function later

File: src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_probe_state.rs [hunk 6]
```diff
@@ -191,84 +198,158 @@ impl HashJoinProbeState {
...
         }
 
+        // Adaptive early filtering.
+        probe_state.num_keys += if let Some(valids) = &valids {
+            (valids.len() - valids.unset_bits()) as u64
+        } else {
+            input_num_rows as u64
+        };
+        let prefer_early_filtering =
+            (probe_state.num_keys_hash_matched as f64) / (probe_state.num_keys as f64) < 0.8;
```
[the remark this answers is not in the document]
Comment (Dousir9) [reply src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_probe_state.rs:263]: 1 / 1，we will try to enable early filtering first.

File: src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_probe_state.rs [hunk 7]
```diff
@@ -191,84 +198,158 @@ impl HashJoinProbeState {
...
+                        )
+                    } else {
+                        // If don't do early filtering, don't use selection.
+                        probe_state.probe_with_selection = false;
+
+                        table.hash_table.probe(&mut probe_state.hashes, valids)
+                    };
+                    probe_state.num_keys_hash_matched += probe_state.selection_count as u64;
+                } else {
+                    // For left join, left single join, full join and left anti join, don't use selection.
```
[the remark this answers is not in the document]
Comment (Dousir9) [reply src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_probe_state.rs:295]: For example, left join need all rows of probe input, so we don't need validity bitmap.

File: src/common/hashtable/src/hashjoin_hashtable.rs (hunk 2 above)
Comment (Dousir9) [reply src/common/hashtable/src/hashjoin_hashtable.rs:176]: > How about
> 
> ```rust
> if let Some(bitmap) = bitmap {
>     if bitmap.unset_bits() == bitmap.len() {
>         hashes.iter_mut().for_each(|hash| {
>             *hash = 0;
>         });
>         return 0;
> }
> ....
> match bitmap {
> 
> }               
> ```

The code in the PR will set bitmap to `None` when bitmap.unset_bits() == 0.

File: src/query/service/src/pipelines/processors/transforms/hash_join/probe_state.rs [hunk 8]
```diff
@@ -23,25 +23,36 @@ use crate::sql::plans::JoinType;
...
     pub(crate) probe_unmatched_indexes: Option<Vec<u32>>,
+    // The `markers` is used for right mark join.
     pub(crate) markers: Option<Vec<u8>>,
-    pub(crate) string_items_buf: Option<Vec<(u64, usize)>>,
+    // Early filtering.
+    pub(crate) selection: Vec<u32>,
+    pub(crate) selection_count: usize,
+    pub(crate) num_keys: u64,
+    pub(crate) num_keys_hash_matched: u64,
+    pub(crate) probe_with_selection: bool,
```
[the remark this answers is not in the document]
Comment (Dousir9) [reply src/query/service/src/pipelines/processors/transforms/hash_join/probe_state.rs:49]: Done.

File: src/common/hashtable/src/traits.rs (hunk 4 above)
Comment (Dousir9) [reply src/common/hashtable/src/traits.rs:515]: Done.

**Comment by BohuTANG** [comment]
This PR passed 3-rounds [wizard double check](https://github.com/BohuTANG/wizard/tree/main/double_check) 🚀 :
```
Starting script execution.
Executing and comparing queries from file: sql/queries.sql
Executing query #1:
-- Query 1 - Inner Join
-- Calculate total sales per customer segment
SELECT c.segment, SUM(s.net_paid) AS total_sales
FROM customers c
         JOIN sales s ON c.customer_id = s.customer_id
GROUP BY c.segment
ORDER BY total_sales DESC
Query #1 results are the same.

Executing query #2:
-- Query 2 - Inner Join
-- Compute average quantity sold per product category
SELECT p.category, AVG(s.quantity) AS avg_quantity
FROM products p
         JOIN sales s ON p.product_id = s.product_id
GROUP BY p.category
ORDER BY avg_quantity DESC
Query #2 results are the same.

Executing query #3:
-- Query 3 - Inner Join with Filtering
-- List top 5 electronics products by total quantity sold
SELECT p.product_name, SUM(s.quantity) AS total_quantity
FROM products p
         JOIN sales s ON p.product_id = s.product_id
WHERE p.category = 'Electronics'
GROUP BY p.product_name
ORDER BY total_quantity DESC
    LIMIT 5
Query #3 results are the same.

Executing query #4:
-- Query 4 - Inner Join
-- Count total sales for each month and year
SELECT dd.year, dd.month, COUNT(*) AS total_sales
FROM date_dim dd
         JOIN sales s ON dd.date_key = s.sale_date
GROUP BY dd.year, dd.month
ORDER BY dd.year, dd.month
    LIMIT 10
Query #4 results are the same.

Executing query #5:
-- Query 5 - Inner Join with Multiple Tables
-- Calculate yearly total sales and average product price
SELECT dd.year, SUM(s.net_paid) AS total_sales, AVG(p.price) AS avg_price
FROM sales s
         JOIN date_dim dd ON s.sale_date = dd.date_key
         JOIN products p ON s.product_id = p.product_id
GROUP BY dd.year
ORDER BY dd.year
Query #5 results are the same.

Executing query #6:
-- Query 6 - Simple Group By with Where Condition
-- Count active customers per segment
SELECT segment, COUNT(*) AS active_customers
FROM customers
WHERE active
GROUP BY segment
ORDER BY active_customers DESC
Query #6 results are the same.

Executing query #7:
-- Query 7 - Inner Join with Aggregation
-- Find average sale amount for each day of the week
SELECT dd.day_of_week, AVG(s.net_paid) AS avg_sale_amount
FROM sales s
         JOIN date_dim dd ON s.sale_date = dd.date_key
GROUP BY dd.day_of_week
ORDER BY avg_sale_amount DESC
Query #7 results are the same.

Executing query #8:
-- Query 8 - Inner Join with Filtering
-- Identify top 3 sales days for furniture category
SELECT dd.date_key, SUM(s.net_paid) AS total_sales
FROM sales s
         JOIN products p ON s.product_id = p.product_id
         JOIN date_dim dd ON s.sale_date = dd.date_key
WHERE p.category = 'Furniture'
GROUP BY dd.date_key
ORDER BY total_sales DESC
    LIMIT 3
Query #8 results are the same.

Executing query #9:
-- Query 9 - Inner Join with Subquery
-- Calculate the contribution percentage of large segment in clothing sales
SELECT SUM(s.net_paid) / (SELECT SUM(net_paid) FROM sales JOIN products ON sales.product_id = products.product_id WHERE products.category = 'Clothing') AS contribution_percentage
FROM sales s
         JOIN products p ON s.product_id = p.product_id
         JOIN customers c ON s.customer_id = c.customer_id
WHERE c.segment = 'Large' AND p.category = 'Clothing'
Query #9 results are the same.

Executing query #10:
-- Query 10 - Inner Join with Window Function
-- Rank customers in each segment by total sales and categorize sales volume
SELECT
    c.customer_id,
    c.customer_name,
    c.segment,
    total_sales,
    avg_product_price,
    max_sale_date,
    rank() OVER (PARTITION BY c.segment ORDER BY total_sales DESC) as sales_rank_in_segment,
        CASE
            WHEN total_sales > 10000 THEN 'High'
            WHEN total_sales BETWEEN 5000 AND 10000 THEN 'Medium'
            ELSE 'Low'
            END as sales_volume_category
FROM
    (SELECT
         s.customer_id,
         SUM(s.net_paid) as total_sales,
         AVG(s.net_paid / s.quantity) as avg_product_price,
         MAX(s.sale_date) as max_sale_date
     FROM
         sales s
     GROUP BY
         s.customer_id) as sales_summary
        JOIN customers c ON sales_summary.customer_id = c.customer_id
WHERE
        c.active = TRUE AND
        sales_summary.total_sales > (SELECT AVG(total_sales) FROM (SELECT customer_id, SUM(net_paid) as total_sales FROM sales GROUP BY customer_id) as average_sales)
ORDER BY
    c.segment, sales_rank_in_segment
    LIMIT 10
Query #10 results are the same.

Executing query #11:
-- Query 11 - Inner Join with Window Function
-- Rank products sold in 2021 by category based on quantity
SELECT
    dd.year,
    p.product_name,
    p.category,
    SUM(s.quantity) as total_sold,
    RANK() OVER (PARTITION BY dd.year, p.category ORDER BY SUM(s.quantity) DESC, p.product_name) as rank_in_category
FROM sales s
         JOIN date_dim dd ON s.sale_date = dd.date_key
         JOIN products p ON s.product_id = p.product_id
GROUP BY dd.year, p.product_name, p.category
HAVING dd.year = 2021
ORDER BY dd.year, p.category, total_sold DESC, p.product_name
    LIMIT 10
Query #11 results are the same.

Executing query #12:
-- Query 12 - Inner Join with Window Function
-- Calculate average sales per category and rank customers by sales
SELECT
    c.customer_name,
    p.category,
    AVG(sales.net_paid) OVER (PARTITION BY p.category) as avg_category_sales,
        s.total_customer_sales,
    RANK() OVER (ORDER BY s.total_customer_sales DESC) as sales_rank,
        CASE
            WHEN s.total_customer_sales > 5000 THEN 'VIP'
            ELSE 'Regular'
            END as customer_status
FROM customers c
         JOIN (
    SELECT customer_id, SUM(net_paid) as total_customer_sales
    FROM sales
    GROUP BY customer_id
) s ON c.customer_id = s.customer_id
         JOIN sales ON sales.customer_id = c.customer_id
         JOIN products p ON sales.product_id = p.product_id
WHERE c.active = TRUE
ORDER BY p.category, sales_rank
    LIMIT 10
Query #12 results are the same.

Executing query #13:
-- Query 13 - Window Function with Common Table Expression (CTE)
-- Compare monthly sales by segment to previous month
WITH monthly_sales AS (
    SELECT
        c.segment,
        EXTRACT(MONTH FROM s.sale_date) as sale_month,
        SUM(s.net_paid) as monthly_sales
    FROM sales s
             JOIN customers c ON s.customer_id = c.customer_id
    GROUP BY c.segment, sale_month
)
SELECT
    segment,
    sale_month,
    monthly_sales,
    LAG(monthly_sales, 1) OVER (PARTITION BY segment ORDER BY sale_month) as previous_month_sales,
            monthly_sales - LAG(monthly_sales, 1) OVER (PARTITION BY segment ORDER BY sale_month) as sales_difference
FROM monthly_sales
ORDER BY segment, sale_month
    LIMIT 10
Query #13 results are the same.

Executing query #14:
-- Query 14 - Left Join
-- Calculate total spending for each customer and segment, including customers with no sales
SELECT c.customer_name, c.segment, COALESCE(SUM(s.net_paid), 0) AS total_spent
FROM customers c
         LEFT JOIN sales s ON c.customer_id = s.customer_id
GROUP BY c.customer_name, c.segment
ORDER BY total_spent DESC, c.customer_name ASC
    LIMIT 10
Query #14 results are the same.

Executing query #15:
-- Query 15 - Right Join
-- Count sales for each product and category, including products with no sales
SELECT p.product_name, p.category, COALESCE(COUNT(s.sale_id), 0) AS sale_count
FROM products p
         RIGHT JOIN sales s ON p.product_id = s.product_id
GROUP BY p.product_name, p.category
ORDER BY p.product_name, sale_count DESC
    LIMIT 10
Query #15 results are the same.

Executing query #16:
-- Query 16 - Full Outer Join
-- List all sales, customers, and products within a date range, including unmatched records
SELECT IFNULL(c.customer_name, 'No Customer') AS customer, IFNULL(p.product_name, 'No Product') AS product, s.sale_date
FROM sales s
         FULL OUTER JOIN customers c ON s.customer_id = c.customer_id
         FULL OUTER JOIN products p ON s.product_id = p.product_id
WHERE s.sale_date BETWEEN '2021-01-01' AND '2021-12-31'
ORDER BY s.sale_date
    LIMIT 10
Query #16 results are the same.

Executing query #17:
-- Query 17 - Inner Join with Multiple Tables
-- Summarize total quantity sold for each customer, product, and year within a specified date range
SELECT c.customer_name, p.product_name, d.year, SUM(s.quantity) AS total_quantity
FROM sales s
         JOIN customers c ON s.customer_id = c.customer_id
         JOIN products p ON s.product_id = p.product_id
         JOIN date_dim d ON d.date_key = s.sale_date
WHERE d.year BETWEEN 2021 AND 2025
GROUP BY c.customer_name, p.product_name, d.year
ORDER BY total_quantity DESC, c.customer_name ASC
    LIMIT 10
Query #17 results are the same.

Executing query #18:
-- Query 18 - Inner Join with Filtering and Aggregation
-- Calculate average spending for active customers, grouped by customer name and segment
SELECT c.customer_name, c.segment, AVG(s.net_paid) AS avg_spent
FROM customers c
         JOIN (
    SELECT customer_id, net_paid
    FROM sales
    WHERE sale_date BETWEEN '2021-01-01' AND '2021-12-31'
) s ON c.customer_id = s.customer_id
WHERE c.active = true
GROUP BY c.customer_name, c.segment, c.customer_id
ORDER BY avg_spent DESC, c.customer_name ASC
    LIMIT 10
Query #18 results are the same.

Executing query #19:
-- Query 19 - Self Join
-- Compare customer sales to find repeat purchases within the same year
SELECT a.customer_id, a.sale_date AS first_sale_date, b.sale_date AS repeat_sale_date
FROM sales a
         JOIN sales b ON a.customer_id = b.customer_id AND EXTRACT(YEAR FROM a.sale_date) = EXTRACT(YEAR FROM b.sale_date) AND a.sale_id <> b.sale_id
ORDER BY a.customer_id, a.sale_date, b.sale_date
    LIMIT 10
Query #19 results are the same.

Executing query #20:
-- Query 20 - Semi Join
-- Select customers who have made at least one purchase
SELECT DISTINCT c.customer_name
FROM customers c
WHERE EXISTS (
    SELECT 1
    FROM sales s
    WHERE c.customer_id = s.customer_id
)
ORDER BY c.customer_name
    LIMIT 10
Query #20 results are the same.

Executing query #21:
-- Query 21 - Right Mark Join
-- List all products and mark whether they have sales
SELECT p.product_name, IF(COUNT(DISTINCT s.sale_id) > 0, 'Has Sales', 'No Sales') AS sale_status
FROM sales s
         RIGHT JOIN products p ON s.product_id = p.product_id
GROUP BY p.product_name
ORDER BY p.product_name, sale_status
    LIMIT 10
Query #21 results are the same.

Script execution completed.
```

**Review by xudong963 — APPROVED** [review]

**Comment by BohuTANG** [comment]
From my tests(many times) with main(v1.2.213), this PR there is no performance improved, and Q18 is more worse, this is one of the result:
image

Test script: https://github.com/BohuTANG/wizard/tree/main/benchsb/bend

**Comment by BohuTANG** [comment]
> From my tests(many times) with main(v1.2.213), this PR there is no performance improved, and Q18 is more worse(this maye), this is one of the result: image
> 
> Test script: https://github.com/BohuTANG/wizard/tree/main/benchsb/bend I think there is something missing, please check @Dousir9 , convert this PR to draft.


Update:
**I made a mistake, the main Q18 is OOM in main but works in the new PR**, let continue this PR.

**Pull request merged**